front/pages/api/w/[wId]/files/[fileId].ts

```diff
@@ -0,0 +1,217 @@
+      },
+    });
+  }
+
+  const user = auth.user();
+  if (!user) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "workspace_user_not_found",
+        message: "Could not find the user of the current session.",
+      },
+    });
+  }
+
+  if (!auth.isUser()) {
+    return apiError(req, res, {
+      status_code: 403,
+      api_error: {
+        type: "workspace_auth_error",
+        message:
+          "Only users of the current workspace can update chat sessions.",
+      },
+    });
+  }
+
+  const { fileId, token } = req.query;
+  if (typeof fileId !== "string" || !isDustFileId(fileId)) {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "Missing fileId query parameter.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET": {
+      const action: Action = validActions.includes(req.query.action as Action)
+        ? (req.query.action as Action)
+        : "download";
+
+      const filePath = makeStorageFilePathForWorkspaceId(owner, fileId);
+
+      if (action === "view") {
+        const stream = await getPrivateUploadBucket().fetchWithStream(filePath);
+        stream.on("error", () => {
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "message_not_found",
+              message: "File not found.",
+            },
+          });
+        });
+
+        stream.pipe(res);
+        return;
+      }
+
+      const fileName = await getFileNameFromFileMetadata(filePath);
+
+      // Redirect to a signed URL.
+      const url = await getPrivateUploadBucket().getSignedUrl(filePath, {
+        // Since we redirect, the use is immediate so expiry can be short.
+        expirationDelay: 10 * 1000,
+        promptSaveAs: fileName ?? `dust_${fileId}`,
+      });
+
+      res.redirect(url);
+      return;
+    }
+
+    case "DELETE": {
+      const filePath = makeStorageFilePathForWorkspaceId(owner, fileId);
+
+      await getPrivateUploadBucket().delete(filePath);
+
+      res.status(204).end();
+      return;
+    }
+
+    case "POST": {
+      if (typeof token !== "string") {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Invalid token.",
+          },
+        });
+      }
+
+      // Ensure that the file token is valid.
+      const fileTokenPayload = decodeFileToken(token);
+
+      if (!fileTokenPayload || fileTokenPayload.fileId !== fileId) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Invalid file id.",
+          },
+        });
+      }
+
+      try {
+        const form = new IncomingForm();
+        const [, files] = await form.parse(req);
```

Comment:
Will this materialize the file in memory? We'd want to stream to GCS instead here no?